src/helpers/serverModuleHelpers.js

```diff
@@ -0,0 +1,122 @@
+const { existsSync } = require('node:fs')
+const { readFile, writeFile, rename, rm } = require('node:fs/promises')
+const { parse, join } = require('node:path')
+
+const { satisfies } = require('semver')
+
+const getAngularJson = require('./getAngularJson')
+const { getProject } = require('./setUpEdgeFunction')
+
+// eslint-disable-next-line no-inline-comments
+const NetlifyServerTsCommonEngine = /* typescript */ `import { CommonEngine } from '@angular/ssr/node'
+import { render } from '@netlify/angular-runtime/common-engine'
+
+const commonEngine = new CommonEngine()
+
+export default async function HttpHandler(request: Request, context: any): Promise<Response> {
+  return await render(commonEngine)
+}
+`
+
+// eslint-disable-next-line no-inline-comments
+const NetlifyServerTsAppEngine = /* typescript */ `import { AngularAppEngine, createRequestHandler } from '@angular/ssr'
+const angularAppEngine = new AngularAppEngine()
+
+// @ts-expect-error - createRequestHandler expects a function with single Request argument and doesn't allow context argument
+export const reqHandler = createRequestHandler(async (request: Request, context: any) => {
+  const result = await angularAppEngine.handle(request, context)
+  return result || new Response('Not found', { status: 404 })
+})
+`
+
+let needSwapping = false
+let serverModuleLocation
+let serverModuleBackupLocation
+
+/**
+ * Inspect content of server module and determine which engine is used
+ * @param {string} serverModuleContents
+ * @returns {'AppEngine' | 'CommonEngine' | undefined}
+ */
+const getUsedEngine = function (serverModuleContents) {
+  if (serverModuleContents.includes('AngularAppEngine') || serverModuleContents.includes('AngularNodeAppEngine')) {
+    return 'AppEngine'
+  }
+
+  if (serverModuleContents.includes('CommonEngine')) {
+    return 'CommonEngine'
+  }
+}
+
+/**
+ * For Angular@19+ we inspect user's server.ts and if it uses express, we swap it out with our own.
+ * We also check wether CommonEngine or AppEngine is used to provide correct replacement preserving
+ * engine of user's choice (CommonEngine is stable, but lacks support for some features, AppEngine is
+ * Developer Preview, but has more features and is easier to integrate with - ultimately choice is up to user
+ * as AppEngine might have breaking changes outside of major version bumps)
+ * @param {Object} obj
+ * @param {string} obj.angularVersion Angular version
+ * @param {string} obj.siteRoot Root directory of an app
+ * @param {(msg: string) => never} obj.failPlugin Function to fail the plugin
+ *
+ * @returns {'AppEngine' | 'CommonEngine' | undefined}
+ */
+const fixServerTs = async function ({ angularVersion, siteRoot, failPlugin }) {
```

Comment:
Quietly swapping out a user's file for another — potentially not running some of their custom code — doesn't seem ideal.

Is there any way we can differentiate between a fully untouched scaffolded default `server.ts` and a customized one, and only replace the former and throw an actionable error on the latter?

😬 I remember seeing that Vercel does this in their Remix adapter: https://github.com/vercel/remix/blob/d8eb9b02a8f803b11272fecbe037725750d0f510/packages/vercel-remix/vite.ts#L100-L109. They have to keep sha checksums of every template that's ever existed. 😰

... Can we chat with the Angular team about ways to make this and the engine check robust instead of relying on brittle hacks?

Reply:
The general gist about silent swapping user code I do agree with, but this is essentially what we do already for Angular 17 and 18 in practice (we just not even use this file, despite users potentially applying customizations there) so this is not something we lose support for as we just never had it - just wanted to make sure this is known.

With above said - as we only care about this module for Angular 19 and later - we probably would be fine with just having ~hashes for the boilerplate that's generated with Angular 19 and treat older as unknown and force users to use one of ours then? Possible try generating current 17 and 18 boiler plate (which would only have CommonEngine) and hash them as CommonEngine ones.

More involved thing would be to look at https://github.com/angular/angular-cli/tree/main/packages/schematics/angular/ssr/files and `server.ts.template` files in those directories and try to hash those (after replacing vars with defaults?) with various git revisions.

The main question is just balance of what we supported historically and wether we want to go with:
 - fail build on unknown `server.ts` forcing user to adjust their code even if they have no actual customizations (maybe they have same code just reformatted due to adjusting prettier/editor settings in which case hashes wouldn't match with known ones)
 - don't fail build and try to be smart - this does lose potential customization but otherwise it ~"just work" and it's inline with current behavior for 17 and 18

> ... Can we chat with the Angular team about ways to make this and the engine check robust instead of relying on brittle hacks?

The difficult part here that the swap kind of has to happen before angular compilation - if we wouldn't want to swap automatically - then we could use some information from Angular Build (if it exist) to inform about build failure message, but we wouldn't be able to use it to automate swapping (unless we hook into compilation itself)